### DISCUSSION TITLE
A CI/CD Pipeline?

### DISCUSSION THREAD

**Discussion by OverloadedOrama** [Orama-Interactive/Pixelorama#252] [Ideas]
I had an idea recently. It is possible to use a CI/CD pipeline with Godot projects? CI could help with code style enforcing, and it could run the code and check if it has any errors. It could also do stuff like, for example, save a .pxo file and then open it, to see if errors are occurring in file saving. If that's even possible, I've never worked with CI before.

And we could use CD to generate executables, perhaps in a new `release` branch or even continuous builds. It could help me automate builds by copying/pasting the brushes/palettes/patterns files next to the executables, and perhaps even use a macOS image to generate `.dmg` files. That would be useful for me because I don't own a Mac, and Godot doesn't have an option to export `.dmg` files if you're not on macOS. Another idea would be to use CD to automate Pixelorama's snap and flatpak packaging. What do you guys think? Is it be a good idea, and is it possible?

EDIT:

Checklist:
**Dev desktop builds worklow.** This gets triggered on pull requests and commits pushed to the `master` branch.
- [x] Generate builds for Windows, Linux and macOS and upload them as artifacts to GitHub Actions.
- [x] Give execute permissions for the Linux and macOS builds.
- [x] Include the `pixelorama_data` folder in Windows, Linux and Mac binaries.
- [x] Export a `.dmg` file instead of a `.zip` for Mac. (Fixed in #319)
- [x] Use WINE and rcedit for Windows builds. Currently done by the workflow, maybe it would be best to be done in the Docker container instead. Related issue - https://github.com/aBARICHELLO/godot-ci/issues/21

**Dev web worklow.** This gets triggered on commits pushed to the `master` branch.
- [x] Generate HTML5 build and deploy it to `gh-pages`, under the "early_access" directory. This makes easier for people to test Pixelorama's new features without having to download it. The page is found in: https://orama-interactive.github.io/Pixelorama/early_access

**Release workflow.** This should get triggered on commits pushed to a `release` branch.
- [x] Generate builds for Windows (64 and 32 bit), Linux (64 and 32 bit) and macOS and upload them to GitHub Releases.
- [x] Generate HTML5 build and deploy it to the root directory of `gh-pages`. The page will be found in: https://orama-interactive.github.io/Pixelorama
- [x] Publish these builds to itch.io via Butler.

**Comment by cfogelklou** [comment]
I might be able to help you guys set up CI/CD using github actions. I've done 3 projects now and am getting the hang of it. (Plus, it's free for open source! :)

**Comment by OverloadedOrama** [comment]
That would be great! I'm still trying to understand how CI/CD works (and what CI actually does, because I feel like I may have misunderstood its purpose) but it should be possible to do with this https://github.com/aBARICHELLO/godot-ci

A potential problem would be having to copy and paste the "pixelorama" folder from the repo to the root directory of the binaries. I'm not sure how easy this is to do.

**Comment by cfogelklou** [comment]
I'll take a look...

**Comment by OverloadedOrama** [comment]
I managed to make a "Test" workflow work in #269. It runs every time a commit is pushed to the `master` branch. I updated the OP with a checklist of things that still need to be done.
